File: app/src/main/java/com/donteco/internetbookstore/models/RepositoryViewModel.java
package com.donteco.internetbookstore.models;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;

import com.donteco.internetbookstore.books.BookInCart;
import com.donteco.internetbookstore.books.CachedBook;
import com.donteco.internetbookstore.books.FullBookInfo;
import com.donteco.internetbookstore.books.ShortenedBookInfo;
import com.donteco.internetbookstore.storage.Repository;

import java.util.List;

public class RepositoryViewModel extends AndroidViewModel {
    private Repository repository;

    private MutableLiveData<String> userInputLiveData;
    private LiveData<List<ShortenedBookInfo>> shortenedInfoBooks;

    public RepositoryViewModel(@NonNull Application application) {
        super(application);

        repository = Repository.getInstance(application);

        userInputLiveData = new MutableLiveData<>();

        shortenedInfoBooks = Transformations.switchMap(userInputLiveData,
                input -> repository.getShortenedBooksInfoByRequest(input));
    }

    public void insertShortenedBooksInfo(List<ShortenedBookInfo> books) {
        repository.insertShortenedBooksInfo(books);
    }

    public void insertCachedBooksInfo(List<CachedBook> books) {
        repository.insertCachedBooksInfo(books);
    }

    public void insertFullBookInfo(FullBookInfo fullBookInfo) {
        repository.insertFullBookInfo(fullBookInfo);
    }

    public LiveData<FullBookInfo> getFullBookInfoById(long id){
        return repository.getFullBookInfoById(id);
    }

    public void setuserInputLiveData(String userInput) {
        userInputLiveData.setValue(userInput);
    }

    public LiveData<List<ShortenedBookInfo>> getShortenedInfoBooks(){
        return shortenedInfoBooks;
    }

    public void insertBookToCart(BookInCart bookInCart){
        repository.insertBookInCartInfo(bookInCart);
    }

    public void updateBookInCart(BookInCart bookInCart){
        repository.updateBookInCartInfo(bookInCart);
    }

    public void deleteBookInCart(BookInCart bookInCart){
        repository.deleteBookInCart(bookInCart);
    }

    public LiveData<BookInCart> getBookInCartById(long id){
        return repository.getBookInCartById(id);
    }

    public LiveData<List<BookInCart>> getCart(){
        return repository.getCart();
    }

}

File: app/src/main/java/com/donteco/internetbookstore/models/RepositoryViewModel.kt
package com.donteco.internetbookstore.models

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Transformations

import com.donteco.internetbookstore.books.BookInCart
import com.donteco.internetbookstore.books.CachedBook
import com.donteco.internetbookstore.books.FullBookInfo
import com.donteco.internetbookstore.books.ShortenedBookInfo
import com.donteco.internetbookstore.storage.Repository

class RepositoryViewModel(application: Application) : AndroidViewModel(application) {
    private val repository: Repository = Repository.getInstance(application)

    private val userInputLiveData: MutableLiveData<String> = MutableLiveData()
    val shortenedInfoBooks: LiveData<List<ShortenedBookInfo>>

    val cart: LiveData<List<BookInCart>>
        get() = repository.cart

    init {
        shortenedInfoBooks = Transformations.switchMap(userInputLiveData) {
            input -> repository.getShortenedBooksInfoByRequest(input) }
    }

    fun insertShortenedBooksInfoAsync(books: List<ShortenedBookInfo>) = repository.insertShortenedBooksInfoAsync(books)

    fun insertCachedBooksInfoAsync(books: List<CachedBook>) = repository.insertCachedBooksInfoAsync(books)

    fun insertFullBookInfoAsync(fullBookInfo: FullBookInfo) = repository.insertFullBookInfoAsync(fullBookInfo)


    fun getFullBookInfoById(id: Long): LiveData<FullBookInfo> {
        return repository.getFullBookInfoById(id)
    }

    fun setUserInputLiveData(userInput: String) {
        userInputLiveData.value = userInput
    }

    fun insertBookToCartAsync(bookInCart: BookInCart) = repository.insertBookInCartInfoAsync(bookInCart)

    fun updateBookInCartAsync(bookInCart: BookInCart) = repository.updateBookInCartInfoAsync(bookInCart)

    fun deleteBookInCartAsync(bookInCart: BookInCart) = repository.deleteBookInCartAsync(bookInCart)

    fun getBookInCartById(id: Long): LiveData<BookInCart> {
        return repository.getBookInCartById(id)
    }
}
